### PR TITLE
NAS-105062 / 21.02 / Allow empty specific attributes for NIC

### PR DESCRIPTION
**Pull request by william-gr**
Ticket:	NAS-105062

### CODE REVIEW

**Review by themylogin — COMMENTED** [review]
It works well for memberless bridge, but how will it work for VLAN without a parent interface or LAGG without a protocol?

**Comment by william-gr** [comment]
@themylogin isn't there code making sure vlan parent iface/lagg proto/etc others are valid?

**Comment by themylogin** [comment]
@william-gr it will pass VLAN without parent https://github.com/freenas/freenas/blob/master/src/middlewared/middlewared/plugins/network.py#L1143 but is it possible to have a VLAN without parent?

**Comment by william-gr** [comment]
Good catch, I will add an else

**Comment by themylogin** [comment]
@william-gr for an empty lagg protocol it will have a poor error message. It will also permit LAGG without interfaces and allow VLAN with empty tag.

**Comment by william-gr** [comment]
Dont think its a poor error message. Its accurate.

There is no null=True for vlan tag

File: src/middlewared/middlewared/plugins/network.py [hunk 1]
```diff
@@ -1110,9 +1110,12 @@ async def _common_validation(self, verrors, schema_name, data, itype, update=Non
             if data['lag_protocol'] not in await self.middleware.call('interface.lag_supported_protocols'):
                 verrors.add(
                     f'{schema_name}.lag_protocol',
-                    f'{platform.system()} does not support LAGG protocol {data["lag_protocol"]}',
+                    f'{platform.system()} does not support LAG protocol {data["lag_protocol"]}',
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/network.py:1113]: if `data["lag_protocol"]` is empty, error message will be `"FreeBSD does not support LAG protocol "`

File: src/middlewared/middlewared/plugins/network.py [hunk 2]
```diff
@@ -1110,9 +1110,12 @@ async def _common_validation(self, verrors, schema_name, data, itype, update=Non
...
                 )
-            for i, member in enumerate(data.get('lag_ports') or []):
+            lag_ports = data.get('lag_ports') or []
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/network.py:1115]: What is the point of `or []` now?

**Review by themylogin — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/plugins/network.py (hunk 1 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/network.py:1113]: Can it be empty if we have enum in the schema?

Comment (themylogin) [reply src/middlewared/middlewared/plugins/network.py:1113]: Good point, sorry

**Review by themylogin — APPROVED** [review]

File: src/middlewared/middlewared/plugins/network.py (hunk 2 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/network.py]: Removed

**Review by themylogin — APPROVED** [review]

**Comment by william-gr** [comment]
Thanks for not allowing me screw up :)

**Comment by william-gr** [comment]
backport

**Pull request merged**